### PR TITLE
feeds: sequence feed resolution fixed and timestamp based heuristic dropped

### PR DESCRIPTION
**Pull request by zelig**
#1248 

- eliminate timestamp based heuristic
- fix sequence async finder bug
- simplify and harden tests
- add test for returned index

### CODE REVIEW

**Comment by agazso** [comment]
It's better now, because it doesn't get stuck at index `1` anymore so the original bug is gone.

I see two new issues when filling up a feed with updates:
- Sometimes the Find feed update does not return. The request is sent, Bee doesn't close the connection but does not return any response.
- Sometimes the next index is out of order. For example after `0f` comes `20` (instead of `10`). Then the subsequent index is `11` but then later `20` is skipped. I also saw out-of-order indexes `40`, `80` and `100`.

This suggests to me that more testing is required.

**Comment by agazso** [comment]
Here is the the tool I use for testing:

https://gist.githubusercontent.com/agazso/f504cfd20c13d723e80e101577b0a6ac/raw/9650dc772d64a20c4bc3990ad71c8ef7673e2556/feed-debug-auto.html

It gets the latest update from the `Find feed update` endpoint, then writes an update using the `swarm-feed-index-next` for calculating the index. Then it waits 1.5 seconds and starts from the beginning. It also keeps track of the index, so if the index is not continuous it will write an error.

Usage: curl the link and save the html file locally. 
Open the file from the browser (or alternatively upload it to swarm and run from there).

For example file:///Users/attila/Projects/bee-js/feed-test-auto.html?bee=http://localhost:1633

You can specify different options in query parameters:
`bee`: the Bee node to connect to
`topic`: change the topic
`privateKey`: change the privateKey
`index`: expect this index to be updated, otherwise error

Running this demonstrates the errors I mentioned above in about 10-20 updates.

File: pkg/feeds/sequence/sequence.go [hunk 1]
```diff
@@ -80,89 +94,110 @@ func NewAsyncFinder(getter storage.Getter, feed *feeds.Feed) feeds.Lookup {
...
-			np := newPath(p.latest.seq)
-			np.level = p.level
-			np.latest.chunk = p.latest.chunk
+			np := p.next()
+			go f.at(ctx, at, np, c, quit)
+		}
+		if p.max < p.min {
+			np := newPath(p.base)
+			np.level = p.max
+			np.max = p.level
```
Comment (zelig) [inline pkg/feeds/sequence/sequence.go:191]: this should be np.max=p.max or np.level

File: pkg/feeds/sequence/sequence.go [hunk 2]
```diff
@@ -80,89 +94,110 @@ func NewAsyncFinder(getter storage.Getter, feed *feeds.Feed) feeds.Lookup {
...
 		p.cancelled = true
 	}
 }
-
+func (p *path) next() *path {
+	return &path{
+		base:   p.index,
+		index:  p.index,
+		max:    p.level,
+		level:  p.level,
```
Comment (zelig) [inline pkg/feeds/sequence/sequence.go:123]: this should be `= p.max`

File: pkg/feeds/sequence/sequence.go [hunk 3]
```diff
@@ -80,89 +94,110 @@ func NewAsyncFinder(getter storage.Getter, feed *feeds.Feed) feeds.Lookup {
...
 		close(p.cancel)
 		p.cancelled = true
 	}
 }
-
+func (p *path) next() *path {
+	return &path{
+		base:   p.index,
+		index:  p.index,
+		max:    p.level,
```
Comment (zelig) [inline pkg/feeds/sequence/sequence.go:122]: this should be `= p.max`

File: pkg/feeds/sequence/sequence.go [hunk 4]
```diff
@@ -80,89 +94,110 @@ func NewAsyncFinder(getter storage.Getter, feed *feeds.Feed) feeds.Lookup {
...
 			}
-			p.close()
-			np := newPath(p.latest.seq)
-			np.level = p.level
-			np.latest.chunk = p.latest.chunk
+			np := p.next()
+			go f.at(ctx, at, np, c, quit)
+		}
+		if p.max < p.min {
+			np := newPath(p.base)
```
Comment (zelig) [inline pkg/feeds/sequence/sequence.go:189]: this is incorrect. we should just use p.Next() and reset max and level to p.level

File: pkg/feeds/testing/lookup.go [hunk 5]
```diff
@@ -73,77 +95,98 @@ func TestFinderBasic(t *testing.T, finderf func(storage.Getter, *feeds.Feed) fee
...
-		})
+		}
 	}
 }
 
 func TestFinderRandomIntervals(t *testing.T, finderf func(storage.Getter, *feeds.Feed) feeds.Lookup, updaterf func(putter storage.Putter, signer crypto.Signer, topic []byte) (feeds.Updater, error)) {
 	for i := 0; i < 5; i++ {
 		t.Run(fmt.Sprintf("random intervals %d", i), func(t *testing.T) {
 			storer := mock.NewStorer()
+			// storer := &Timeout{mock.NewStorer()}
```
Comment (zelig) [inline pkg/feeds/testing/lookup.go:189]: remove

File: pkg/feeds/sequence/sequence.go (hunk 1 above)
Comment (zelig) [reply pkg/feeds/sequence/sequence.go:191]: ^

File: pkg/feeds/sequence/sequence.go (hunk 2 above)
Comment (zelig) [reply pkg/feeds/sequence/sequence.go:123]: ^

Comment (zelig) [reply pkg/feeds/sequence/sequence.go:123]: ^

**Review by zelig — COMMENTED** [review]
LGTM

**Comment by Eknir** [comment]
This PR is ready for review, but not connected to an issue. The issue which belongs to the PR is still in the needs confirmation pipeline. Please properly manage your issues.

**Comment by Eknir** [comment]
@zelig , is this ready for review?

**Comment by zelig** [comment]
@Eknir now it is. you can  trust my use of labels :)

**Comment by agazso** [comment]
`index` query parameter is not implemented, it would be good to remove from the documentation (or implement it):

https://docs.ethswarm.org/api/#tag/Feed/paths/~1feeds~1{owner}~1{topic}/get

File: pkg/feeds/testing/lookup.go [hunk 6]
```diff
@@ -17,10 +19,30 @@ import (
...
+// Get overrides the mock storer and introduces latency
+func (t *Timeout) Get(ctx context.Context, mode storage.ModeGet, addr swarm.Address) (swarm.Chunk, error) {
+	ch, err := t.Storer.Get(ctx, mode, addr)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			time.Sleep(searchTimeout)
+		}
+		return ch, err
+	}
+	time.Sleep(time.Duration(rand.Intn(10)) * time.Millisecond)
```
Comment (acud) [inline pkg/feeds/testing/lookup.go:40]: I would increase this significantly, perhaps to 100 or more milliseconds delay. Also, this assures that chunks which are always returned before the chunks which are not found. Theoretically OK, but does not give us any test coverage that can simulate jitter

**Review by acud — CHANGES_REQUESTED** [review]

File: pkg/feeds/testing/lookup.go (hunk 6 above)
Comment (zelig) [reply pkg/feeds/testing/lookup.go:40]: I think it is ok, i tested and it does what it is meant to  do. 
increasing the time 10 fold would just slow down tests

File: pkg/feeds/sequence/sequence.go [hunk 7]
```diff
@@ -80,132 +99,157 @@ func NewAsyncFinder(getter storage.Getter, feed *feeds.Feed) feeds.Lookup {
...
-			np.latest.chunk = p.latest.chunk
-			go f.at(ctx, at, np, c, quit)
+			go f.at(ctx, at, 0, i.next(), c, quit)
+		}
+		// inconsistent feed, retry
+		if i.notFound < i.found.level {
+			go f.at(ctx, at, i.found.level, i.retry(), c, quit)
 		}
 	}
 	return nil, nil, nil, nil
```
[the remark this answers is not in the document]
Comment (zelig) [reply pkg/feeds/sequence/sequence.go:198]: ok I will do this

**Review by acud — APPROVED** [review]

**Comment by acud** [comment]
Let's address further changes in subsequent PRs

**Pull request merged**